packages/@o3r/design/src/core/design-token/renderers/design-token-style.renderer.ts

```diff
@@ -22,9 +22,62 @@ export const computeFileToUpdatePath = (root = process.cwd(), defaultFile = 'sty
  * Compare Token variable by name
  * @param a first token variable
  * @param b second token variable
+ * @deprecated use {@link getTokenSorterByName} instead. Will be removed in v13.
  */
 export const compareVariableByName = (a: DesignTokenVariableStructure, b: DesignTokenVariableStructure): number => a.getKey().localeCompare(b.getKey());
 
+/**
+ * Compare Token variable by name
+ * @param _variableSet Complete set of the parsed Design Token
+ */
+export const getTokenSorterByName: TokenListTransform = (_variableSet) => (tokens) => tokens.sort((a, b) => a.getKey().localeCompare(b.getKey()));
+
+
+/**
+ * Compare Token variable by reference to write theme in reference order
+ * @param variableSet Complete set of the parsed Design Token
+ */
+export const getTokenSorterByRef: TokenListTransform = (variableSet) => {
+  /**
+   * Get the maximum number of reference levels for a Design Token, following its references
+   * @param token Design Token
+   * @param level Value of the current level
+   * @param mem Memory of the visited node
+   */
+  const getReferenceLevel = (token: DesignTokenVariableStructure, level = 0, mem: DesignTokenVariableStructure[] = []): number => {
+    const children = token.getReferencesNode(variableSet);
+    if (children.length === 0 || mem.includes(token)) {
+      return level;
+    } else {
+      level++;
+      mem = [...mem, token];
+      return Math.max(...children.map((child) => getReferenceLevel(child, level, mem)));
+    }
+  };
+
+  return (tokens) => {
+    const limit = Math.max(...tokens.map((t) => getReferenceLevel(t)));
+    let sortToken = [...tokens];
+    for (let i = 0; i < limit + 1; i++) {
+      let hasChanged = false;
+      sortToken = sortToken.reduce((acc, token, idx) => {
+        const firstRef = sortToken.findIndex((t) => t.getReferences(variableSet).includes(token.tokenReferenceName));
+        if (firstRef > -1 && firstRef < idx) {
+          acc.splice(idx, 1);
+          acc.splice(firstRef, 0, token);
+          hasChanged = true;
+        }
+        return acc;
+      }, sortToken);
```

Comment:
Are you sure reduce is working with a copy of sortToken? It seems rather dangerous to modify what you are iterating on.

Reply:
![image](https://github.com/user-attachments/assets/0ea88269-fec8-4b0d-9588-4de71be9686e)
In this case, it stops iterating after the 2 second element because idx > myTestArray.length
![image](https://github.com/user-attachments/assets/e15b8891-5faa-4d9b-bb56-cbb84435af40)

Reply:
especially with splice in the loop :)

Reply:
You actually have an issue in your example because you change the number of items in the original array (you could also give an example where you desynchronize the index of the loop by adding more item before the current index and get an issue as well).

In the current code there is:
- no array length change
- no change of number of items in left and right part between the index.

An example of very similar case is a standard function to revert an array:

```javascript
const testArray = ['a', 'b', 'c', 'd'];

testArray.reduce((acc, key, index) => {
    console.log(acc, key, index);
    acc.splice(index, 1);
    acc.unshift(key);
    return acc;
}, testArray);

console.log('result: ', testArray);
```

which output:
```
['a', 'b', 'c', 'd'] 'a' 0
['a', 'b', 'c', 'd'] 'b' 1
['b', 'a', 'c', 'd'] 'c' 2
['c', 'b', 'a', 'd'] 'd' 3
result: ['d', 'c', 'b', 'a']
```

Reply:
This is why I mentionned that in your particular case it works (see my remark "I know it should not be the case because you remove one element and add one before the idx, but it is not easy too read and it is risky for maintenance"), however I would not consider it a good practice as we could easily mess up the code while maintaining it.

Reply:
My bad I got only the first message on VsCode so I replied with the first one as context only.
In anycase, the code has changed, you can see if it fills better your expectations